src/main/java/org/logstash/dissect/Dissector.java

```diff
@@ -0,0 +1,180 @@
+            if (list.isEmpty()) {
+                if (delimiter.size() > 0)
+                    initialOffset = delimiter.size();
+            } else {
+                temp.setPrevious(delimiter);
+                list.get(fieldIndex - 1).setNext(delimiter);
+            }
+            list.add(fieldIndex, temp);
+        }
+        return list;
+    }
+
+    private void createLastField(List<FieldDelimiterHolder> list) {
+        FieldDelimiterHolder last;
+        // the fields List has the last element removed to allow for a
+        // different way to set the value
+        int lastFieldIndex = list.size() - 1;
+        if ((list.size() - 1) > -1) {
+            last = list.remove(lastFieldIndex);
+            lastField = FieldFactory.create(last.name, last.previous, last.next);
+            saveableFields.add(lastField);
+        }
+    }
+
+    private void createFieldList(List<FieldDelimiterHolder> list) {
+        for (FieldDelimiterHolder holder : list) {
+            Field field = FieldFactory.create(holder.name, holder.previous, holder.next);
+            fields.add(field);
+            if (field.saveable()) {
+                saveableFields.add(field);
+            }
+        }
+
+        // the saveableFields List is fields List minus the Skip fields
+        // sorted so AppendFields are last
+        Collections.sort(saveableFields, new FieldComparator());
+    }
+
+    public int dissect(byte[] source, Map<String, Object> keyValueMap) {
+        final Map<Field, ValueRef> fieldValueRefMap = createFieldValueRefMap();
+        int pos = dissectValues(source, fieldValueRefMap);
+        ValueResolver resolver = new ValueResolver(source, fieldValueRefMap);
+
+        for (Field field : saveableFields) {
+            field.append(keyValueMap, resolver);
+        }
+        return pos;
+    }
+
+    public int dissect(byte[] source, Event event) {
+        final Map<Field, ValueRef> fieldValueRefMap = createFieldValueRefMap();
+        int pos = dissectValues(source, fieldValueRefMap);
+        ValueResolver resolver = new ValueResolver(source, fieldValueRefMap);
+
+        for (Field field : saveableFields) {
+            field.append(event, resolver);
+        }
+        return pos;
+    }
+
+    /*
+        An IdentityHashMap is needed here because fields can have the same name and ordinal
+        but differ in the identities of next and previous Delimiter instances.
+        So we really compare by identity.
+        We could consider adding an instanceId int property when creating Fields and
+        this would mean that one Field is equal to another Field if they have the same instanceId.
+     */
+    private Map<Field, ValueRef> createFieldValueRefMap() {
+        Map<Field, ValueRef> map = new IdentityHashMap<>(fields.size() + 1);
+        map.put(NormalField.MISSING, new ValueRef(0, 0));
+        return map;
+    }
+
+    private int dissectValues(byte[] source, Map<Field, ValueRef> fieldValueRefMap) {
```

Comment:
I think some comments here and a few other methods will really help the reader and the next person maintaining this code...